sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamZetaSqlCatalog.java

```diff
@@ -276,6 +285,49 @@ private void addWindowTvfs() {
             null));
   }
 
+  private void addUdfsFromSchema() {
+    for (String functionName : calciteSchema.getFunctionNames()) {
+      Collection<org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.schema.Function>
+          functions = calciteSchema.getFunctions(functionName);
+      if (functions.size() != 1) {
+        throw new IllegalArgumentException(
+            String.format(
+                "Expected exactly 1 definition for function '%s', but found %d.",
```

Comment:
Is this an assert or a Beam ZetaSQL limitation? Is the case where we get multiple values back is when the user supplied multiple variants of the same function? If so: `Beam ZetaSQL supports exactly 1...`

Users are going to eventually ask for multiple variants of the same UDF for different types, it appears that is already supported by Calcite and ZetaSQL.

Reply:
It's a Beam ZetaSQL limitation. I filed a JIRA for it: BEAM-12073